echopype/utils/prov.py

```diff
@@ -30,20 +35,63 @@ def echopype_prov_attrs(process_type: ProcessType) -> Dict[str, str]:
     return prov_dict
 
 
+def _source_files(paths: Union[str, Path, List[Any], Tuple[Any], np.ndarray]):
+    """
+    Create sanitized list of string paths from heterogeneous path inputs.
+
+    Parameters
+    ----------
+    paths : Union[str, Path, List[str, Path], Tuple[str, Path], np.ndarray]
+        File paths as either a single path string or pathlib Path,
+        a sequence (tuple, list or np.ndarray) of strings or pathlib Paths,
+        or a mixed sequence that may contain another sequence as an element.
+
+    Returns
+    -------
+    paths_list : List[str]
+        List of file paths. Empty list if no source path element was parsed successfully.
+    """
+    sequence_types = (list, tuple, np.ndarray)
+    if isinstance(paths, (str, Path)):
+        return [str(paths)]
+    elif isinstance(paths, sequence_types):
+        paths_list = []
+        for p in paths:
+            if isinstance(p, (str, Path)):
+                paths_list.append(str(p))
+            elif isinstance(p, sequence_types):
+                paths_list += [str(pp) for pp in p if isinstance(pp, (str, Path))]
+            else:
+                logger.warning(
+                    f"Unrecognized file path element type, path element will not be written to (meta)source_file provenance attribute. {p}"  # noqa
+                )
+        return paths_list
+    else:
+        logger.warning(
+            f"Unrecognized file path element type, path element will not be written to (meta)source_file provenance attribute. {paths}"  # noqa
+        )
+        return []
+
+
 def source_files_vars(
-    source_paths: Union[str, List[Any]], meta_source_paths: Union[str, List[Any]] = None
+    source_paths: Union[str, Path, List[Any], Tuple[Any], np.ndarray],
+    meta_source_paths: Union[None, str, Path, List[Any], Tuple[Any], np.ndarray] = None,
```

Comment:
The `List[Any], Tuple[Any]` should actually be `List[str, Path], Tuple[str, Path]` as what you had in the docstring?

Reply:
Heh, good eye. Initially I tried having `List[str, Path]` and `Tuple[str, Path]` in the argument signature. But it looks like type hinting doesn't allow that, b/c I got an error. So I went with `List[Any]` and `Tuple[Any]` but kept the more explicit version in the docstring in order to, well, be more explicit. I wonder if using `Union[str, Path]` instead of `str, Path` would do the trick. If it does, it'll make for an accurate but pretty long and hard-to-digest type specification. 

I also didn't try to see if `np.ndarray` can be followed by a sequence element type specification. I did some research just now, and it turns out that numpy added a typing library last year or so, [numpy.typing](https://numpy.org/devdocs/reference/typing.html). It includes a better replacement for `np.ndarray`: `numpy.typing.NDArray`. It also has `numpy.typing.ArrayLike`, which *may* be a more compact way to handle `List`, `Tuple` and `np.ndarray`. I'll look into it. It'll be great if that's the case!

Reply:
After much R&D, I submitted a "perfect" (IMHO) implementation of the type hints using aliases, and it turns out it [fails with Python 3.8 ](https://github.com/OSOceanAcoustics/echopype/actions/runs/3744903528/jobs/6358775827#step:14:36) :sob: . I'll revisit it later.